sdk/identity/Azure.Identity/src/ProcessRunner.cs

```diff
@@ -0,0 +1,138 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Azure.Identity
+{
+#pragma warning disable CA1001 // Types that own disposable fields should be disposable. Disposing of _process and _ctRegistration / _timeoutCtRegistration fields from outside may result in _tcs being incomplete or process handle leak.
+    internal sealed class ProcessRunner
+#pragma warning restore CA1001 // Types that own disposable fields should be disposable
+    {
+        private readonly IProcess _process;
+        private readonly TaskCompletionSource<string> _tcs;
+        private readonly CancellationToken _cancellationToken;
+        private readonly CancellationToken _timeoutCancellationToken;
+        private readonly CancellationTokenSource _timeoutCts;
+        private readonly CancellationTokenRegistration _ctRegistration;
+        private readonly CancellationTokenRegistration _timeoutCtRegistration;
+
+        public ProcessRunner(IProcess process, TimeSpan timeout, CancellationToken cancellationToken)
+        {
+            _process = process;
+            _cancellationToken = cancellationToken;
+            _tcs = new TaskCompletionSource<string>();
+
+            _ctRegistration = cancellationToken.Register(QueueHandleCancel);
+            if (timeout.TotalMilliseconds >= 0)
+            {
+                _timeoutCts = new CancellationTokenSource();
```

Comment:
Would CancellationTokenSource.Combine() with CancellationTokenSource.CancellAfter be appropriate here?

Reply:
It may be valuable for the code that has created `ProcessRunner` to check if it has been canceled by `timeout` or by `cancellationToken`

Reply:
@AlexanderSher we do need to be able to distinguish when the authentication was cancelled by the user versus by us. If cancelled by us we should raise a AuthenticationFailedException rather than OperationCancelledException or TaskCancelledException. But using CancellationTokenSource.Combine does not prevent us from doing this, as we can query each token to see if it was cancelled.

Reply:
> If cancelled by us

I.e. by timeout?

> But using CancellationTokenSource.Combine does not prevent us from doing this, as we can query each token to see if it was cancelled.

This will save one `CancellationTokenRegistration` field, but make code more error prone. Since combined `CancellationTokenSource` can be canceled by both user `CancellationToken` and timeout, we should always first check that user `CancellationToken`. With current implementation, order doesn't matter.

Reply:
@AlexanderSher yes "cancelled by us" in this context means we cancelled due to a timeout. We need to be able to distinguish this case from the CancellationToken passed in by the user being cancelled. That's the requirement here, as to the most straight forward way to accomplish this I would defer to what you and @pakrym think is best.

Reply:
Fixed